src/Features/Core/Portable/SignatureHelp/SignatureHelpService.cs

```diff
@@ -0,0 +1,108 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Extensions;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.SignatureHelp;
+
+/// <summary>
+/// A service that is used to determine the appropriate signature help for a position in a document.
+/// </summary>
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+[Export(typeof(ISignatureHelpService)), Shared]
+internal sealed class SignatureHelpService([ImportMany] IEnumerable<Lazy<ISignatureHelpProvider, OrderableLanguageMetadata>> allProviders)
+    : ISignatureHelpService
+{
+    private readonly ConcurrentDictionary<string, ImmutableArray<ISignatureHelpProvider>> _providersByLanguage = [];
+    private readonly IEnumerable<Lazy<ISignatureHelpProvider, OrderableLanguageMetadata>> _allProviders = allProviders;
+
+    private ImmutableArray<ISignatureHelpProvider> GetProviders(string language)
+    {
+        return _providersByLanguage.GetOrAdd(language, language
+            => _allProviders.Where(p => p.Metadata.Language == language)
+                .SelectAsArray(p => p.Value));
+    }
+
+    /// <summary>
+    /// Gets the <see cref="ISignatureHelpProvider"/> and <see cref="SignatureHelpItems"/> associated with
+    /// the position in the document.
+    /// </summary>
+    public Task<(ISignatureHelpProvider? provider, SignatureHelpItems? bestItems)> GetSignatureHelpAsync(
+        Document document,
+        int position,
+        SignatureHelpTriggerInfo triggerInfo,
+        SignatureHelpOptions options,
+        CancellationToken cancellationToken)
+    {
+        return GetSignatureHelpAsync(
+            GetProviders(document.Project.Language),
+            document,
+            position,
+            triggerInfo,
+            options,
+            cancellationToken);
+    }
+
+    /// <summary>
+    /// Gets the <see cref="ISignatureHelpProvider"/> and <see cref="SignatureHelpItems"/> associated with
+    /// the position in the document.
+    /// </summary>
+    public async Task<(ISignatureHelpProvider? provider, SignatureHelpItems? bestItems)> GetSignatureHelpAsync(
+        ImmutableArray<ISignatureHelpProvider> providers,
+        Document document,
+        int position,
+        SignatureHelpTriggerInfo triggerInfo,
+        SignatureHelpOptions options,
+        CancellationToken cancellationToken)
+    {
+        var extensionManager = document.Project.Solution.Services.GetRequiredService<IExtensionManager>();
+
+        ISignatureHelpProvider? bestProvider = null;
+        SignatureHelpItems? bestItems = null;
```

Comment:
i'm assuming this was a move, and didn't change.

Reply:
Ah, yeah I meant more generally the logic moved but not verbatim. On a suggestion from DavidB it uses the extension manager when invoking providers as well as inlines the `IsBetter` method.